Add option to sort discussions by file name

johnybx opened the pull request:
Hi,
Currently we have discussion tree as list of discussions but when discussion tree is opened it is little bit hard to figure out where the discussions actually are located ( in which files ). You can jump from discussion to reviewer and back but I wanted to have better overview so I was playing around with discussion tree and added option to render file tree which is similar to diffview.nvim file panel. I am still experimenting with it and probably would make sense to write some tests for that :thinking: . Would you agree to add such feature ?

harrisoncramer commented:
Yeah I like this idea, the comments by file name looks quite nice. 

My only concern is that we are introducing a lot of code to maintain in the `discussions` module that may be hard to debug, especially since it's all currently untested. I'm not opposed to doing this, but would want to make sure we're being thoughtful about changing this module.

johnybx commented:
Hi,
yes, I definitely agree that discussions module is getting quite big. I pushed initial update how we can split discussions and if you agree with this I would further also split signs and diagnostics logic to separate files. I also added scripts to be able to run lua tests for modules which depends on other plugins and neovim specific functions ( for example now you can use `vim.inspect` in test ). Currently you should be able to run script which will initialize everything for you (if needed) and run tests like this ( dots should be green :sweat_smile: ):

```bash
❯ ./lua-test.sh
●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●
34 successes / 0 failures / 0 errors / 0 pending : 0.0311 seconds
```

I want to add more tests for discussions because currently I added just basic tests for initializing unlinked discussion nodes.
We will also need to update github workflow because currently it uses `lunarmodules/busted@v2.2.0` but that is not sufficient for running tests - we will need to update steps so that during tests we have available `nvim`, `luarocks` and `git` commands ( didn't have time to check this yet ).

I am not sure where do you want to document testing and development as such. Have you considered something like https://generator.contributing.md/?

Please let me know if you agree with this approach or if you prefer something else :pray:.

johnybx commented:
I updated github workflow and now it runs tests with stable and nightly neovim. I noticed that tests are in 2 workflows - do you want to keep separate `lua-tests.yaml` or should I move updated workflow to `lua.yaml` ?

harrisoncramer commented:
> I updated github workflow and now it runs tests with stable and nightly neovim. I noticed that tests are in 2 workflows - do you want to keep separate `lua-tests.yaml` or should I move updated workflow to `lua.yaml` ?

Yeah we should just have a single flow for Lua tests. Thanks for doing this! I'd been planning to set up running tests through Neovim but hadn't figured out how to do it yet.

harrisoncramer commented:
Hey, thank you for this MR. It's pretty big so I'll try to review it this weekend. And yes I've thought about adding contributing and will work that into my next MR which is also focused mostly on testing.

harrisoncramer commented:
Whenever you send a `note_id` it should be an integer and you are sending a string now. Both the `send_deletion` and `send_edits` functions aren't working for me for either the new tree type or the old tree type.

johnybx commented on tests/spec/discussions_tree_spec.lua:
```diff
@@ -0,0 +1,167 @@
+      assert.are.same(#nodes[2].__children[2].__children, 1)
+      assert.are.same(nodes[2].__children[2].__children[1].text, "Response to the unlinked note")
+      assert.are.same(nodes[2].__children[2].__children[1].type, "note_body")
+      assert.spy(spy_time_since).was.called_with("2023-11-20T20:15:49.648Z")
```
these were different but I see your point, I will update it to have it more distinguishable.
